api/v1alpha1/ratelimit_types.go

```diff
@@ -108,49 +108,25 @@ type SourceMatch struct {
 }
 
 // HeaderMatch defines the match attributes within the HTTP Headers of the request.
-type HeaderMatch struct { // TODO: zhaohuabing this type could be replaced with a general purpose StringMatch type.
-	// Type specifies how to match against the value of the header.
-	//
-	// +optional
-	// +kubebuilder:default=Exact
-	Type *HeaderMatchType `json:"type,omitempty"`
-
+type HeaderMatch struct {
 	// Name of the HTTP header.
 	// +kubebuilder:validation:MinLength=1
 	// +kubebuilder:validation:MaxLength=256
 	Name string `json:"name"`
 
-	// Value within the HTTP header. Due to the
-	// case-insensitivity of header names, "foo" and "Foo" are considered equivalent.
+	// Distinct matches any and all possible unique values encountered in the
+	// specified HTTP Header.
+	// Note that each unique value will receive its own rate limit bucket.
+	// Only one of Distinct or Match can be set.
+	Distinct *bool `json:"distinct,omitempty"`
+
+	// Match specifies how to match against the value of the header.
 	// Do not set this field when Type="Distinct", implying matching on any/all unique
 	// values within the header.
-	//
-	// +optional
-	// +kubebuilder:validation:MaxLength=1024
-	Value *string `json:"value,omitempty"`
+	// Only one of Distinct or Match can be set.
+	Match *StringMatch `json:"match,omitempty"`
```

Comment:
@arkodg @zirain @Xunzhuo  I plan to refactor rate limit HeaderMatch to use the shared string match. If you agree this is the right way, I'll go ahead.

Reply:
we should keep header matches aligned to upstream as much as possible.
for this case we defined in this api because we introduced a new match type called `distinct` that doesnt exist upstream
https://github.com/kubernetes-sigs/gateway-api/blob/5fd8bec3bedc7c325bf2227da48d041a735b9b98/apis/v1beta1/httproute_types.go#L98
https://github.com/kubernetes-sigs/gateway-api/blob/5fd8bec3bedc7c325bf2227da48d041a735b9b98/apis/v1/httproute_types.go#L439

Also although sprinkling different match types affects project mantainence, it benefits end users because for each different feature we are adding a different set of kubebuilder annotations for the fields, so we providing proactive feedback on the type of matches that we support.

tldr, if the match types are identical, it makes sense to consolidate, else imo we should wait until they reach parity

Reply:
IMO, the Distinct Match type only matters when it comes to the Rate Limit. So, perhaps we could consider renaming HeaderMatch to RateLimitHeaderMatch. Let's leave it as is for now and see how we want to deal with the string match when we have new APIs in the mix.